subsys/mgmt/mcumgr/smp_bt.c

```diff
@@ -89,62 +106,70 @@ static struct conn_param_data *get_conn_param_data(const struct bt_conn *conn)
 }
 
 /* Sets connection parameters for a given conn. */
-static void set_conn_latency(struct bt_conn *conn, bool low_latency)
+static void conn_params_set(struct bt_conn *conn, bool smp_params)
 {
-	struct bt_le_conn_param params;
-	struct conn_param_data *cpd;
-	struct bt_conn_info info;
 	int ret = 0;
+	const struct bt_le_conn_param *params;
+	struct conn_param_data *cpd;
 
 	cpd = get_conn_param_data(conn);
+	params = smp_params ? SMP_BT_CONN_PARAMS : DEFAULT_CONN_PARAMS;
 
-	ret = bt_conn_get_info(conn, &info);
-	__ASSERT_NO_MSG(!ret);
-
-	if ((low_latency && (info.le.latency == 0)) ||
-	    ((!low_latency) && (info.le.latency == DEFAULT_LATENCY))) {
-		/* Already updated. */
-		return;
+	ret = bt_conn_le_param_update(conn, params);
+	if (ret && (ret != -EALREADY)) {
+		/* Try again to avoid stucking with incorrect connection parameters. */
+		(void)k_work_reschedule(&cpd->ework, K_MSEC(RESTORE_RETRY_TIME));
+	} else {
+		(void)k_work_cancel_delayable(&cpd->ework);
 	}
 
-	params.interval_min = info.le.interval;
-	params.interval_max = info.le.interval;
-	params.latency = (low_latency) ? (0) : (DEFAULT_LATENCY);
-	params.timeout = info.le.timeout;
-
-	ret = bt_conn_le_param_update(cpd->conn, &params);
-	if (ret && (ret != -EALREADY)) {
-		if (!low_latency) {
-			/* Try again to avoid stucking in low latency. */
-			(void)k_work_reschedule(&cpd->dwork, K_MSEC(RESTORE_RETRY_TIME));
+	if (!ret) {
+		if (smp_params) {
+			cpd->state |= SMP_CONN_PARAMS_REQUESTED;
+			cpd->state &= ~DEFAULT_CONN_PARAMS_REQUESTED;
+		} else {
+			cpd->state |= DEFAULT_CONN_PARAMS_REQUESTED;
+			cpd->state &= ~SMP_CONN_PARAMS_REQUESTED;
 		}
 	}
 }
 
 
-/* Work handler function for restoring the default latency for the connection. */
-static void restore_default_latency(struct k_work *work)
+/* Work handler function for restoring the default connection parameters for the connection. */
+static void on_pref_conn_params_restore(struct k_work *work)
 {
 	struct conn_param_data *cpd;
 
 	cpd = CONTAINER_OF(work, struct conn_param_data, dwork);
 
-	if (cpd->latency_state & CONN_LOW_LATENCY_REQUIRED) {
-		cpd->latency_state &= ~CONN_LOW_LATENCY_REQUIRED;
-		(void)k_work_reschedule(&cpd->dwork, K_MSEC(RESTORE_TIME));
-	} else {
-		set_conn_latency(cpd->conn, false);
+	/* Unmark that the SMP Connection Parameters are necessary. */
+	cpd->state &= ~SMP_CONN_PARAMS_REQUIRED;
+
+	if (cpd->state & SMP_CONN_PARAMS_ENABLED) {
```

Comment:
There could be a race condition where conn params update callback for SMP parameters is called after the timeout. It could result in being stuck with SMP params. Discussed offline.

Reply:
I think we should remove the check.